apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+            if (CurrentActiveSessions < MaximumActiveSessions
+                && s_waitQueue.TryDequeue(out waitingSessionRequest))
+            {
+                waitingSessionRequest.SetResult(0);
+            }
+        }
+
+        private static void LogSessionsInUse()
+        {
+            Logger.LogPerformanceCounter("Session.ActiveCount", () => s_sessionsInUse.Count);
+        }
+
+        /// <summary>
+        /// Releases a session back into the pool, possibly causing another entry to be evicted if the maximum pool size has been
+        /// reached.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <param name="client"></param>
+        /// <returns></returns>
+        public static void ReleaseToPool(this SpannerClient client, Session session )
+        {
+            session.ThrowIfNull(nameof(session));
+
+            SessionPoolKey result;
+            if (s_sessionsInUse.TryRemove(session, out result))
+            {
+                LogSessionsInUse();
+                SignalAnyWaitingRequests();
+                SessionPoolImpl targetPool = GlobalPool.GetOrAdd(result,
+                                             key => new SessionPoolImpl(key));
+
+                SpannerClient evictionClient = result.Client;
+                Session evictionSession = null;
+
+                //Figure out if we want to pool this released session.
+                lock (s_priorityListSync)
+                {
+                    targetPool.ReleaseSessionToPool(client, session);
+                    if (CurrentPooledSessions > MaximumPooledSessions)
+                    {
+                        var evictionPool = PriorityList.First().Value;
+                        evictionClient = evictionPool.Key.Client;
+                        evictionSession = evictionPool.AcquireEvictionCandidate();
+                        ReSortMru(evictionPool);
+                    }
+                    ReSortMru(targetPool);
+                }
+                if (evictionSession != null)
+                {
+                    Task.Run(() => EvictSession(evictionClient, evictionSession));
+                }
+            }
+        }
+
+        private static async Task EvictSession(SpannerClient evictionClient, Session evictionSession)
+        {
+            await evictionSession.RemoveFromTransactionPool().ConfigureAwait(false);
+            await evictionClient.DeleteSessionAsync(evictionSession.SessionName);
+        }
+
+        /// <summary>
+        /// Called to close a session and release it without putting it back into the pool.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <returns></returns>
+        public static async Task Close(this Session session)
+        {
+            session.ThrowIfNull(nameof(session));
+
+            SessionPoolKey result;
+            if (s_sessionsInUse.TryRemove(session, out result))
```

Comment:
no -- this is used in cases where we do not want to push the session back into the pool.
we're basically just deleting it and updating our state of its use.  sessionpoolimpl only has state on pooled sessions.